OCI layout documentation (#574)
buildpacks/docs: content/docs/app-developer-guide/export-to-oci-layout.md

+++
title="Export to OCI layout format"
weight=3
summary="Learn how to export your application image to disk in OCI layout format"
+++

<div class="quote mb-4">
    The OCI Image Layout is the directory structure for OCI content-addressable blobs and location-addressable references.

>     <div class="author">See the <a href="https://github.com/opencontainers/image-spec/blob/main/image-layout.md">specification</a></div>

Reviewer: ```suggestion
    <div class="author">See the <a href="https://github.com/opencontainers/image-spec/blob/main/image-layout.md">specification.</a></div>
```

</div>

Exporting to OCI layout format is an **experimental** feature available on pack since version X.Y.Z

> 

Reviewer 2: I think we need a table somewhere comparing the `cnb` layout with the `oci` layout.  The first question I'd ask as a user is "Which layout should I choose?", and such a table would contain enough "Why?" to answer which layout to choose.

### 1. Enable experimental feature

> Verify your pack version is equal or greater than X.Y.Z

Reviewer 2: ```suggestion
Verify your pack version is equal or greater than 0.29.0
```

```bash
pack version
```

Enable the experimental features on pack

```bash
pack config experimental true
```

You can confirm everything is fine, checking the `config.toml` file in your `PACK_HOME` installation folder, for example:

```bash
cat ~/.pack/config.toml
experimental = true

> layout-repo-dir = "<$HOME>/.pack/layout-repo"

Reviewer 2: Does `pack config experimental true` add the "layout-repo-dir" to `config.toml`?

Author: Yes, that's correct

```

The configuration shows the experimental mode was **enabled** and a local directory to save images on disk was configured to path `<$HOME>/.pack/layout-repo`.

### 2. Build the app

> If you haven't already, please follow the steps to [build an app](/docs/app-developer-guide/build-an-app), once you managed to build a sample application you can try exporting the sample application to disk in OCI layout format. 

Reviewer 2: ```suggestion
Please first follow the steps to [build an app](/docs/app-developer-guide/build-an-app), once you have successfully built an application you can export the sample application to disk in OCI layout format. 
```

We try and stay away from words like "managed", or terminology that suggests the process might be difficult.

The OCI layout feature must be enabled using the convention `oci:<path/to/save/image>` in the `<image-name>` parameter when invoking `pack build`.

For example:

```bash
pack build oci:sample-app --path samples/apps/java-maven --builder cnbs/sample-builder:bionic
```

> It will save the image in a folder `./sample-app` created in your current directory.

Reviewer 2: Why do we need `layout-repo-dir` if the OCI layout is saved to the PWD?

Author: We are using the `layout-repo-dir` to save the `run-image` on disk because actually invoking the lifecycle.  When lifecycle is configured to use OCI layout it is expecting to have all the require inputs (like the run-image) on disk,  because that should be transparent for users, we are saving those images in that particular folder.

```bash
tree sample-app

sample-app
├── blobs
│ └── sha256
│     ├── 141bfb0cd434d425bc70edb9e56ea11d07aed76450eb0e73e6110645f251a8d3
│     ├── 2fa192256ce255c6ea6c1296eadfe2feba8094f40e6aa85e699645caca2e85d8
│     ├── 5a44e4f7b58d74fe6f92dd7028075c91191128d1e2e7f39846fe061a9a98836e
│     ├── 72d9f18d70f395ff9bfae4d193077ccea3ca583e3da3dd66f5c84520c0100727
│     ├── 827746ec7ba80f4e4811b6c9195b6f810fbc2d58a6c9cc337bf0305791f24e97
│     ├── ad13830c92258c952f25d561d8bf7d9eb58b8a3003960db1502cbda8239130b5
│     ├── b97b58b190d5f731c879b0f7446a2bd554863b51851e03757199c74dd922ce61
│     ├── c44222730efa142cd5bedc0babf82a9a07d325494be7f5c3cfde56f43166b65f
│     ├── e1048fb89c3194a1f0542c0847aa086a7034dd7867c48fe8c93675cf36f90610
│     ├── f0a30c5bc44742065b1b4ffa95271a39994f05ba7a03dd7e7143d1d3e45fa0b1
│     └── f9d6350d0c44c0e7165a522155f53181ce8c163a6b8ead1f6baea22d1a8d8a78
├── index.json
└── oci-layout  

3 directories, 13 files
```
If you are interested on keep playing with the image in  OCI layout format, one tool you can take a look is [umoci](https://umo.ci/) it can help you to create a 

> [runtime bundler](https://github.com/opencontainers/runtime-spec) that can be executed with another tool like [runc](https://github.com/opencontainers/runc)

Reviewer 2: There a number of tools here that we might need to explain to end-users.  I *think* we're using `umoci` because it is a strict-OCI compliant tool?  That assembles an image.  I _presume_ that docker can also execute a strict-OCI image?  If so, should we prefer to not introduce a new runtime there?

Author: Yeah, I had some doubts about including those tool, maybe we can suppose users know what they can do with the OCI layout format

---

## Extra configuration

### Skip saving your run-image layers on disk

Before using this option we suggest to remove your local layout directory (the one configured in your pack config.toml with the key `layout-repo-dir`) and 

> your application image folder (if you are planning to use the same folder), the reason for this is pack doesn't remove the blobs saved in the `layout-repo-dir` if you use the `--sparse` flag 

Reviewer 2: Here it becomes more unclear why `layout-repo-dir` was necessary above.

├── index.json
└── oci-layout

3 directories, 10 files
```

As you can see, there are 3 missing files at `sample-app/blobs/sha256` folder, the missing 3 blobs are the blobs from the 
`run-image` that were not downloaded but if you check your config file you'll notice you have the same number of layers as 
when you export the full image.

## Implementation notes

### Media Types

According to the OCI specification, the [compatibles media types](https://github.com/opencontainers/image-spec/blob/main/media-types.md#compatibility-matrix) for the index.json files must be:

- `application/vnd.oci.image.index.v1+json` or
- `application/vnd.docker.distribution.manifest.list.v2+json` 

If you are trying to use the lifecycle directly without using `pack` to export your image, take on consideration that tools like:

[skopeo](https://github.com/containers/skopeo)
```bash
skopeo copy -a docker://<your-image> <dest>
```
It will give you `application/vnd.oci.image.index.v1+json` media type, which is currently working

But [crane](https://github.com/google/go-containerregistry/tree/main/cmd/crane) 

```bash
crane pull <your-image> <dest> --format=oci
```

> It will give you `application/vnd.docker.distribution.manifest.list.v2+json`, which will fail

Reviewer 2: Do these things pass/fail due to the state of _our_ current implementation?  Or because of a limitation in `crane`, or because the operation violates part of the OCI spec? Or for some other reason?

Reviewer: agree, helpful to add a little more detail here

Author: It fails because of the state or our current implementation, it should be improve on new versions of the feature. I will add it

Author: Let me explain this part a little better so you can guide me on how to document it. 

According to the [RFC](https://github.com/buildpacks/rfcs/blob/main/text/0119-export-to-oci.md#how-it-works) if you want to export your application image on disk, you need to mount to the lifecycle a *local storage* with the images require by the build process, for example, the `run-image`. This storage must contain images in OCI layout format and name resolution will follow some rules defined also in the RFC. 

The documented part was based on a feedback provided by @natalieparellano [here](https://github.com/buildpacks/rfcs/pull/203#discussion_r1092449172), which is basically the following scenario: Suppose you want to execute the lifecycle directly without using pack, for example, a platform implementor. They need to create this local storage and save everything there before executing the build, right?, well the problem is: *If they save in the local storage a manifest list* and then they execute the lifecycle, right now, we are not smart enough to detect which manifest we need to pick to execute the build. 

That is the actual comment we want to document so users can take it into consideration, now I am not sure this is the right place to do it, or maybe I didn't write it correctly, so any advice will be appreciate!
